lib/spack/llnl/util/filesystem.py

```diff
@@ -466,29 +471,109 @@ def fix_darwin_install_name(path):
                         stdout=subprocess.PIPE).communicate()[0]
                     break
 
-# Utilities for libraries
 
+def find(root, files, recurse=True):
+    """Search for ``files`` starting from the ``root`` directory.
 
-class LibraryList(collections.Sequence):
-    """Sequence of absolute paths to libraries
+    Like GNU/BSD find but written entirely in Python.
 
-    Provides a few convenience methods to manipulate library paths and get
-    commonly used compiler flags or names
+    Examples:
+
+    .. code-block:: console
+
+       $ find /usr -name python
+
+    is equivalent to:
+
+    .. code-block:: python
+
+       >>> find('/usr', 'python')
+
+    .. code-block:: console
+
+       $ find /usr/local/bin -maxdepth 1 -name python
+
+    is equivalent to:
+
+    .. code-block:: python
+
+       >>> find('/usr/local/bin', 'python', recurse=False)
+
+    Accepts any glob characters accepted by fnmatch:
+
+    =======  ==================================
+    Pattern  Meaning
+    =======  ==================================
+    *        matches everything
+    ?        matches any single character
+    [seq]    matches any character in `seq`
+    [!seq]   matches any character not in `seq`
+    =======  ==================================
+
+    :param str root: The root directory to start searching from
+    :param files: Library name(s) to search for
+    :type files: str or collections.Sequence
+    :param bool recurse: if False search only root folder,
+        if True descends top-down from the root
+
+    :returns: The files that have been found
+    :rtype: list
+    """
+    if isinstance(files, str):
+        files = [files]
+
+    if recurse:
+        return _find_recursive(root, files)
+    else:
+        return _find_non_recursive(root, files)
+
+
+def _find_recursive(root, search_files):
+    found_files = []
+
+    for path, _, list_files in os.walk(root):
+        for search_file in search_files:
+            for list_file in list_files:
+                if fnmatch.fnmatch(list_file, search_file):
+                    found_files.append(join_path(path, list_file))
+
+    return found_files
+
+
+def _find_non_recursive(root, search_files):
+    found_files = []
+
+    for list_file in os.listdir(root):
+        for search_file in search_files:
+            if fnmatch.fnmatch(list_file, search_file):
+                found_files.append(join_path(root, list_file))
+
+    return found_files
+
+
+# Utilities for libraries and headers
+
+
+class FileList(collections.Sequence):
+    """Sequence of absolute paths to files.
+
+    Provides a few convenience methods to manipulate file paths.
     """
 
-    def __init__(self, libraries):
-        self.libraries = list(libraries)
+    def __init__(self, files):
+        self.files = list(dedupe(files))
```

Comment:
@alalazo One change I made was to remove duplicates from the list when creating the object. Is there any reason this would be a problem?